File: medium/lengthOfLongestSubstr.go
package main

import "fmt"

/**
 * https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/
 */
func main(){
	ret := lengthOfLongestSubstring("pwwkew")
	fmt.Printf("ret  %v", ret)
}

func lengthOfLongestSubstring(s string) int {
	if s == "" {
		return 0
	}

	b, ret := 0, 0
	chars := make(map[rune] int)

	for k,v := range s {
		i, ok := chars[v]

		if ok {
			if i + 1 > b {
				b = i + 1
			}
		}

		chars[v] = k

		if k - b + 1 > ret {
			ret = k - b + 1
		}
	}

	return ret
}

File: medium/lengthOfLongestSubstr.php
<?php

class Solution {

    /**
     * https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        $ret = 0;

        if (!empty($s)) {      
            $b = 0; //初始化滑动窗口的开始位置
            $chars = [];

            for ($i = 0; $i < strlen($s); $i++) {
                if (array_key_exists($s[$i], $chars)) {
                    $b = max($chars[$s[$i]] + 1, $b);
                }

                $chars[$s[$i]] = $i;
                $ret = max($ret, $i - $b + 1);
            }
        }

        return $ret;
    }
}
